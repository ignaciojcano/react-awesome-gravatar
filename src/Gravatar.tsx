import * as React from 'react';
import { getGravatarUrl, GravatarOptions } from './lib';

export interface GravatarProps extends React.ImgHTMLAttributes<any> {
  email: string;
  options?: GravatarOptions;
  component?: any;
}

export default class Gravatar extends React.PureComponent<GravatarProps, {}> {
  static defaultProps: Partial<GravatarProps> = {
    component: (props: any) => {
      return (
        <img { ...props }/>
      );
    },
  };

  buildComponentProps() {
    const componentProps = {
      ...this.props,
      src: getGravatarUrl(this.props.email, this.props.options),
    };
    delete componentProps.email;
    delete componentProps.options;
    delete componentProps.component;
    return componentProps;
  }

  render(): JSX.Element {
    return this.props.component(this.buildComponentProps());
  }
}
