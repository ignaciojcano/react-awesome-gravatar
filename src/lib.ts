import { stringify } from 'query-string';
import md5 from 'md5';

export interface GravatarOptions {
  size?: number;
  default?: '404'|'mp'|'identicon'|'monsterid'|'wavatar'|'retro'|'robohash'|'blank';
  defaultUrl?: string;
  forcedefault?: 'y';
  rating?: 'g'|'pg'|'r'|'x';
}

export const getGravatarUrl = (email: string, options: GravatarOptions = {}): string => {
  const query: any = { ...options };
  if (options.defaultUrl) {
    query.default = options.defaultUrl;
  }
  const emailHash = md5(email);
  return `https://www.gravatar.com/avatar/${emailHash}?${stringify(query)}`;
};
