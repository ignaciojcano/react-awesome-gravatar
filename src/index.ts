import Gravatar from './Gravatar';
export { getGravatarUrl, GravatarOptions } from './lib';

export default Gravatar;
