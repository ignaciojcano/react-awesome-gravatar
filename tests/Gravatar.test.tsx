import * as React from 'react';
import { create } from 'react-test-renderer';
import Gravatar from '../src/Gravatar';
import { GravatarOptions } from '../src/lib';

describe('getGravatarUrl', () => {
  const email:string = 'example@example.com';

  it('Renders the default img tag with the correct url', () => {
    const component = create(
      <Gravatar email={email} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Renders the img tag with the correct url and options', () => {
    const options: GravatarOptions = {
      default: 'monsterid',
      size: 250,
    };
    const component = create(
      <Gravatar email={email} options={options} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Renders a custom component with the correct url passed in through props', () => {
    const options: GravatarOptions = {
      default: 'monsterid',
      size: 250,
    };
    const textComponent = (props: any) => {
      const { src, ...rest } = props;

      return (
        <text value={src} {...rest}/>
      );
    };
    const component = create(
      <Gravatar email={email} options={options} component={textComponent} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
