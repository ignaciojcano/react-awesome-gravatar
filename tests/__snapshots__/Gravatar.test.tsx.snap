// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGravatarUrl Renders the default img tag with the correct url 1`] = `
<img
  src="https://www.gravatar.com/avatar/23463b99b62a72f26ed677cc556c44e8?"
/>
`;

exports[`getGravatarUrl Renders the img tag with the correct url and options 1`] = `
<img
  src="https://www.gravatar.com/avatar/23463b99b62a72f26ed677cc556c44e8?default=monsterid&size=250"
/>
`;
